scripts/prepdocs/files.py

```diff
@@ -0,0 +1,749 @@
+                batches.append(batch)
+                batch = []
+                batch_token_length = 0
+
+        if len(batch) > 0:
+            batches.append(batch)
+
+        return batches
+
+    async def create_embedding_batch(self, texts: List[str]) -> List[List[float]]:
+        batches = self.split_text_into_batches(texts)
+        for batch in batches:
+            async for attempt in AsyncRetrying(
+                retry=retry_if_exception_type(openai.error.RateLimitError),
+                wait=wait_random_exponential(min=15, max=60),
+                stop=stop_after_attempt(15),
+                before_sleep=self.before_retry_sleep,
+            ):
+                with attempt:
+                    emb_args = await self.create_embedding_arguments()
+                    emb_response = await openai.Embedding.acreate(**emb_args, input=batch)
+                    return [data["embedding"] for data in emb_response["data"]]
+
+    async def create_embedding_single(self, text: str) -> List[float]:
+        async for attempt in AsyncRetrying(
+            retry=retry_if_exception_type(openai.error.RateLimitError),
+            wait=wait_random_exponential(min=15, max=60),
+            stop=stop_after_attempt(15),
+            before_sleep=self.before_retry_sleep,
+        ):
+            with attempt:
+                emb_args = await self.create_embedding_arguments()
+                emb_response = await openai.Embedding.acreate(**emb_args, input=text)
+                return emb_response["data"][0]["embedding"]
+
+    async def create_embeddings(self, texts: List[str]) -> List[List[float]]:
+        if not self.disable_batch and self.open_ai_model_name in OpenAIEmbeddings.SUPPORTED_BATCH_AOAI_MODEL:
+            return await self.create_embedding_batch(texts)
+
+        return [await self.create_embedding_single(text) for text in texts]
+
+
+class AzureOpenAIEmbeddingService(OpenAIEmbeddings):
+    def __init__(
+        self,
+        open_ai_service: str,
+        open_ai_deployment: str,
+        open_ai_model_name: str,
+        credential: Union[AsyncTokenCredential, AzureKeyCredential],
+        disable_batch: bool = False,
+        verbose: bool = False,
+    ):
+        self.open_ai_service = open_ai_service
+        self.open_ai_deployment = open_ai_deployment
+        self.open_ai_model_name = open_ai_model_name
+        self.credential = credential
+        self.disable_batch = disable_batch
+        self.cached_token = None
+        self.verbose = verbose
+
+    async def create_embedding_arguments(self) -> dict[str, Any]:
+        return {
+            "model": self.open_ai_model_name,
+            "deployment_id": self.open_ai_deployment,
+            "api_type": self.get_api_type(),
+            "api_key": await self.wrap_credential(),
+            "api_version": "2023-05-15",
+        }
+
+    def get_api_type(self) -> str:
+        return "azure_ad" if self.credential is AsyncTokenCredential else "azure"
```

Comment:
hm. shouldnt this be isinstance? (vs. is?)
maybe I'm blanking on how Python works. Same with other is check.

Reply:
you are right :)